src/main/java/com/metamx/metrics/JvmMonitor.java

```diff
@@ -206,14 +115,165 @@ public boolean doMonitor(ServiceEmitter emitter)
       emitter.emit(builder.build("jvm/bufferpool/count", pool.getCount()));
     }
+  }
 
-    return true;
+  private void emitGcMetrics(ServiceEmitter emitter)
+  {
+    gcCounters.emit(emitter, dimensions);
   }
 
-  private void emitGcMetrics(GcCounters gcCounters, ServiceEmitter emitter)
+  /*
+   * The following GC-related code is partially based on
+   * https://github.com/aragozin/jvm-tools/blob/e0e37692648951440aa1a4ea5046261cb360df70/
+   * sjk-core/src/main/java/org/gridkit/jvmtool/PerfCounterGcCpuUsageMonitor.java
+   */
+  private class GcCounters
+  {
+    private List<GcGeneration> generations = new ArrayList<>();
+
+    GcCounters()
+    {
+      long currentProcessId = SigarUtil.getCurrentProcessId();
+      // connect to itself
+      JStatData jStatData = JStatData.connect(currentProcessId);
+      Map<String, JStatData.Counter<?>> jStatCounters = jStatData.getAllCounters();
+
+      generations.add(new GcGeneration(jStatCounters, 0, "young"));
+      generations.add(new GcGeneration(jStatCounters, 1, "old"));
+      // Removed in Java 8 but still actual for previous Java versions
+      if (jStatCounters.containsKey("sun.gc.generation.2.name")) {
+        generations.add(new GcGeneration(jStatCounters, 2, "perm"));
+      }
+    }
+
+    void emit(ServiceEmitter emitter, Map<String, String[]> dimensions)
+    {
+      for (GcGeneration generation : generations) {
+        final ServiceMetricEvent.Builder builder = new ServiceMetricEvent.Builder();
+        MonitorUtils.addDimensionsToBuilder(builder, dimensions);
+        generation.emit(emitter, builder);
+      }
+    }
+  }
+
+  private class GcGeneration
   {
-    final ServiceMetricEvent.Builder builder = new ServiceMetricEvent.Builder();
-    MonitorUtils.addDimensionsToBuilder(builder, dimensions);
-    gcCounters.emitCounters(emitter, builder);
+    private String name;
+    private GcGenerationCollector collector = null;
+    private List<GcGenerationSpace> spaces = new ArrayList<>();
+
+    GcGeneration(Map<String, JStatData.Counter<?>> jStatCounters, long genIndex, String name)
+    {
+      this.name = name.toLowerCase();
+
+      long spacesCount = ((JStatData.LongCounter) jStatCounters.get(String.format("sun.gc.generation.%d.spaces", genIndex))).getLong();
+      for (long spaceIndex = 0; spaceIndex < spacesCount; spaceIndex++) {
+        spaces.add(new GcGenerationSpace(jStatCounters, genIndex, spaceIndex));
+      }
+
+      if (jStatCounters.containsKey(String.format("sun.gc.collector.%d.name", genIndex))) {
+        collector = new GcGenerationCollector(jStatCounters, genIndex);
+      }
+    }
+
+    void emit(ServiceEmitter emitter, ServiceMetricEvent.Builder builder)
+    {
+      builder.setDimension("gcGen", name);
+
+      if (collector != null) {
+        collector.emit(emitter, builder);
+      }
+
+      for (GcGenerationSpace space: spaces) {
+        space.emit(emitter, builder);
```

Comment:
Since each `GcGenerationSpace` fills the same dimension I think that's ok to pass the same builder, because the same dimension will be replaced.